tests/test_fem.py

```diff
@@ -343,3 +344,138 @@ def test_linear_corotated_sphere_fall_implicit_fem_sap_coupler(fem_material_line
         assert_allclose(y_scale, 0.2, atol=1e-3), f"Entity {entity.uid} Y scale {y_scale} is not close to 0.2."
         # The Z scale is expected to be more squashed due to gravity
         assert_allclose(z_scale, 0.2, atol=2e-3), f"Entity {entity.uid} Z scale {z_scale} is not close to 0.2."
+
+
+def test_box_hard_vertex_constraint(show_viewer):
+    """
+        Test if a box with hard vertex constraints has those vertices fixed, 
+        and that updating and removing constraints works correctly.
+    """
+    scene = gs.Scene(
+        sim_options=gs.options.SimOptions(
+            dt=1e-3,
+            substeps=2,
+        ),
+        fem_options=gs.options.FEMOptions(
+            use_implicit_solver=False,
+            gravity=(0.0, 0.0, -9.81),
+        ),
+        show_viewer=show_viewer,
+        show_FPS=False,
+    )
+
+    box = scene.add_entity(
+        morph=gs.morphs.Box(
+            size=(0.1, 0.1, 0.1),
+            pos=(0.0, 0.0, 0.5),
+        ),
+        material=gs.materials.FEM.Elastic(),
+    )
+    vertex_indices = [0, 3]
+    initial_target_positions = box.init_positions[vertex_indices]
+    
+    scene.build()
+
+    if show_viewer:
+        scene.draw_debug_spheres(
+            poss=initial_target_positions,
+            radius=0.02,
+            color=(1, 0, 1, 0.8)
+        )
+
+    box.set_vertex_constraints(
+        vertex_indices=vertex_indices,
+        target_positions=initial_target_positions
+    )
+
+    for _ in range(100):
+        scene.step()
+
+    positions = box.get_state().pos[0][vertex_indices]
+    assert_allclose(
+        positions, initial_target_positions, tol=0.0
+    ), "Vertices should stay at initial target positions with hard constraints"
+    new_target_positions = initial_target_positions + gs.tensor(
+        [[0.1, 0.1, 0.1], [0.1, 0.1, 0.1]], 
+    )
+    box.update_constraint_targets(
+        vertex_indices=vertex_indices,
+        target_positions=new_target_positions
+    )
+
+    for _ in range(100):
+        scene.step()
+
+    positions_after_update = box.get_state().pos[0][vertex_indices]
+    assert_allclose(
+        positions_after_update,
+        new_target_positions,
+        tol=0.0
+    ), "Vertices should be at new target positions after updating constraints"
+
+    box.remove_vertex_constraints(vertex_indices)
+
+    for _ in range(100):
+        scene.step()
+
+    positions_after_removal = box.get_state().pos[0][vertex_indices]
+
+    with np.testing.assert_raises(AssertionError):
+        assert_allclose(
+            positions_after_removal,
+            new_target_positions,
+            tol=1e-3
+        ), "Vertices should have moved after removing constraints"
+
+def test_box_soft_vertex_constraint(show_viewer):
+    """Test if a box with strong soft vertex constraints has those vertices near."""
+    scene = gs.Scene(
+        sim_options=gs.options.SimOptions(
+            dt=1e-4,
+            substeps=10,
```

Comment:
Are you sure this many substeps are really necessary? I would rather use `substeps=1` and `dt=1e-5` if this is really necessary.